OpenRA.Mods.Common/EditorBrushes/EditorCopyPasteBrush.cs

```diff
@@ -182,33 +112,81 @@ sealed class CopyPasteEditorAction : IEditorAction
-		readonly CellLayer<byte> mapHeight;
-		readonly CellLayer<ResourceTile> mapResources;
-
-		readonly Queue<UndoCopyPaste> undoCopyPastes = new();
-		readonly Queue<EditorActorPreview> removedActors = new();
-		readonly Queue<EditorActorPreview> addedActorPreviews = new();
-
-		public CopyPasteEditorAction(MapCopyFilters copyFilters, Map map,
-			Dictionary<CPos, (TerrainTile Tile, ResourceTile Resource, byte Height)> tiles, Dictionary<string, ActorReference> previews,
-			EditorActorLayer editorLayer, CellRegion dest)
+		readonly IResourceLayer resourceLayer;
+		readonly EditorActorLayer editorActorLayer;
+		readonly EditorClipboard clipboard;
+		readonly EditorClipboard undoClipboard;
+		readonly CPos pastePosition;
+		readonly Map map;
+
+		public CopyPasteEditorAction(
+			MapCopyFilters copyFilters,
+			IResourceLayer resourceLayer,
+			CPos pastePosition,
+			Map map,
+			EditorClipboard clipboard,
+			EditorActorLayer editorActorLayer)
 		{
 			this.copyFilters = copyFilters;
-			this.tiles = tiles;
-			this.previews = previews;
-			this.editorLayer = editorLayer;
-			this.dest = dest;
+			this.resourceLayer = resourceLayer;
+			this.clipboard = clipboard;
+			this.pastePosition = pastePosition;
+			this.editorActorLayer = editorActorLayer;
+			this.map = map;
 
-			mapTiles = map.Tiles;
-			mapHeight = map.Height;
-			mapResources = map.Resources;
+			undoClipboard = CopySelectionContents();
 
-			Text = TranslationProvider.GetString(CopiedTiles, Translation.Arguments("amount", tiles.Count));
+			Text = TranslationProvider.GetString(CopiedTiles, Translation.Arguments("amount", clipboard.Tiles.Count));
+		}
+
+		/// <summary>
+		/// TODO: This is pretty much repeated in MapEditorSelectionLogic.
+		/// </summary>
+		/// <returns>Clipboard containing map contents for this region.</returns>
+		EditorClipboard CopySelectionContents()
+		{
+			var selectionSize = clipboard.CellRegion.BottomRight - clipboard.CellRegion.TopLeft;
+			var source = new CellCoordsRegion(pastePosition, pastePosition + selectionSize);
+			var selection = new CellRegion(map.Grid.Type, pastePosition, pastePosition + selectionSize);
+
+			var mapTiles = map.Tiles;
+			var mapHeight = map.Height;
+			var mapResources = map.Resources;
+
+			var previews = new Dictionary<string, EditorActorPreview>();
+			var tiles = new Dictionary<CPos, ClipboardTile>();
+
+			foreach (var cell in source)
+			{
+				if (!mapTiles.Contains(cell))
+					continue;
+
+				var offset = new CVec(clipboard.CellRegion.TopLeft.X, clipboard.CellRegion.TopLeft.Y);
+
+				var resourceLayerContents = resourceLayer.GetResource(cell);
+
+				tiles.Add(cell, new ClipboardTile
+				{
+					TerrainTile = mapTiles[cell],
+					ResourceTile = mapResources[cell],
+					ResourceLayerContents = resourceLayerContents,
+					Height = mapHeight[cell]
+				});
+
+				if (copyFilters.HasFlag(MapCopyFilters.Actors))
+				{
+					var regionActors = selection.SelectMany(editorActorLayer.PreviewsAt).Distinct().ToList();
```

Comment:
```suggestion
					var regionActors = selection.SelectMany(editorActorLayer.PreviewsAt).ToHashSet();
```

Same here and bellow